src/analysis/normed_space/affine_isometry.lean

```diff
@@ -0,0 +1,441 @@
+/-
+Copyright (c) 2021 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+import analysis.normed_space.add_torsor
+import analysis.normed_space.linear_isometry
+import linear_algebra.affine_space.affine_subspace
+
+/-!
+# Affine isometries
+
+In this file we define `affine_isometry 𝕜 P P₂` (notation: `P →ᵃⁱ[𝕜] P₂`) to be a affine isometric
+embedding of normed add-torsors `P` into `P₂` over normed `𝕜`-spaces and `affine_isometry_equiv`
+(notation: `P ≃ᵃⁱ[𝕜] P₂`) to be a affine isometric equivalence between `P` and `P₂`.
+
+We also prove basic lemmas and provide convenience constructors.  The choice of these lemmas and
+constructors is closely modelled on those for the `linear_isometry` and `affine_map` theories.
+
+Since many elementary properties don't require `∥x∥ = 0 → x = 0` we initially set up the theory for
+`semi_normed_add_torsor` and specialize to `normed_add_torsor` only when needed.
+-/
+open function set
+
+variables (𝕜 : Type*) {V V₁ V₂ V₃ V₄ : Type*} {P₁ : Type*} (P P₂ : Type*) {P₃ P₄ : Type*}
+    [normed_field 𝕜]
+  [semi_normed_group V] [normed_group V₁] [semi_normed_group V₂] [semi_normed_group V₃]
+    [semi_normed_group V₄]
+  [semi_normed_space 𝕜 V] [normed_space 𝕜 V₁] [semi_normed_space 𝕜 V₂] [semi_normed_space 𝕜 V₃]
+    [semi_normed_space 𝕜 V₄]
+  [pseudo_metric_space P] [metric_space P₁] [pseudo_metric_space P₂] [pseudo_metric_space P₃]
+    [pseudo_metric_space P₄]
+  [semi_normed_add_torsor V P] [normed_add_torsor V₁ P₁] [semi_normed_add_torsor V₂ P₂]
+    [semi_normed_add_torsor V₃ P₃] [semi_normed_add_torsor V₄ P₄]
+
+include V V₂
+
+/-- An `𝕜`-affine isometric embedding of one normed add-torsor over a normed `𝕜`-space into
+another. -/
+structure affine_isometry extends P →ᵃ[𝕜] P₂ :=
+(norm_map : ∀ x : V, ∥linear x∥ = ∥x∥)
+
+omit V V₂
+variables {𝕜 P P₂}
+
+notation P ` →ᵃⁱ[`:25 𝕜:25 `] `:0 P₂:0 := affine_isometry 𝕜 P P₂
+
+namespace affine_isometry
+variables (f : P →ᵃⁱ[𝕜] P₂)
+
+/-- The underlying linear map of an affine isometry is in fact a linear isometry. -/
+protected def linear_isometry : V →ₗᵢ[𝕜] V₂ :=
+{ norm_map' := f.norm_map,
+  .. f.linear }
+
+@[simp] lemma linear_eq_linear_isometry : f.linear = f.linear_isometry.to_linear_map :=
+by { ext, refl }
+
+include V V₂
+instance : has_coe_to_fun (P →ᵃⁱ[𝕜] P₂) := ⟨_, λ f, f.to_fun⟩
+omit V V₂
+
```

Comment:
Can you add an `initialize_simps_projection` line here like we do for `linear_map` and `linear_equiv`?

Reply:
Can you explain more? (or feel free to push to the branch yourself.). I don't know how this works.

Reply:
I mean copy this line with the obvious replacement:

https://github.com/leanprover-community/mathlib/blob/fdeb0646f567cbc8e0c56cf74069a60f1126be10/src/algebra/module/linear_map.lean#L81

Reply:
And these lines which are less obvious:

https://github.com/leanprover-community/mathlib/blob/fdeb0646f567cbc8e0c56cf74069a60f1126be10/src/algebra/module/linear_map.lean#L503-L506

Reply:
Am I doing this right?  When I put
```lean
initialize_simps_projections affine_isometry (to_fun → apply)
```
I get the error
`Failed to find constructor to_fun in structure affine_isometry.`

Reply:
Oh, this is because this is a new structure, and most maps are old structures. This works:
```
initialize_simps_projections affine_isometry (to_affine_map_to_fun → apply)
```

Reply:
This doesn't seem to work as I expected though. Let's not bother after all unless @fpvandoorn has an opinion.

Reply:
You can mimic `local_homeomorph` to configure simps for new structures that extend other structures.
For this type, I did it here (but only added `simps` to 1 definition): https://github.com/leanprover-community/mathlib/compare/affine-isometry-1-simps

Feel free to merge that and this branch (and feel free not to do it).

Reply:
This was quite instructive, thank you, although I decided not to merge the branch (see [Zulip](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.238660)).